File: Misc DMOJ/dph/src/dph.java
import java.util.*;
import java.io.*;
public class dph {
	/* 
	 * Copy-pasting code is NOT cool! Please do not copy and paste my code as a submission to DMOJ.
	 * github.com/jdabtieu/competitive-programming
	 */
	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer in = new StringTokenizer(br.readLine());
		
		int rows = Integer.parseInt(in.nextToken()), cols = Integer.parseInt(in.nextToken());
		int[][] grid = new int[cols][rows];
		
		for (int i = 0; i < rows; i++) {
			in = new StringTokenizer(br.readLine());
			String data = in.nextToken();
			for (int j = 0; j < data.length(); j++) {
				if (data.charAt(j) == '#') grid[j][i] = -1;
			}
		}
		
		grid[0][0] = 1;
		
		for (int i = 0; i < cols; i++) {
			for (int j = 0; j < rows; j++) {
				if (grid[i][j] != -1) {
					long process = 0;
					if (i != 0 && grid[i-1][j] != -1) process += grid[i-1][j];
					if (j != 0 && grid[i][j-1] != -1) process += grid[i][j-1];
					grid[i][j] += (int) (process % 1000000007);
				}
			}
		}
		System.out.println(grid[cols-1][rows-1]);
	}

}

File: Atcoder DP/dph/src/dph.cpp
#include <bits/stdc++.h>
#pragma GCC optimize ("Ofast")

using namespace std;
/*
 * Copy-pasting code is NOT cool! Please do not copy and paste my code as a submission to DMOJ.
 * github.com/jdabtieu/competitive-programming
 */
int g[1001][1001];
const int mod = 1000000007;
int main() {
    int r, c;
    cin >> r >> c;
    for (int i = 0; i < r; i++) {
        string s;
        cin >> s;
        for (int j = 0; j < c; j++) {
            if (s[j] == '#') g[i][j] = -1;
        }
    }
    g[0][0] = 1;
    for (int i = 0; i < r; i++) {
        for (int j = 0; j < c; j++) {
            if (g[i][j] == -1) continue;
            if (g[i+1][j] != -1) g[i+1][j] = (g[i+1][j] + g[i][j]) % mod;
            if (g[i][j+1] != -1) g[i][j+1] = (g[i][j+1] + g[i][j]) % mod;
        }
    }
    cout << g[r-1][c-1] << endl;
}
